consensus/src/chained_bft/liveness/pacemaker_timeout_manager.rs

```diff
@@ -109,38 +110,47 @@ impl PacemakerTimeoutManager {
     /// Returns the highest round PacemakerTimeoutCertificate from a map of author to
     /// timeout messages or None if there are not enough timeout messages available.
     /// A PacemakerTimeoutCertificate is made of the N highest timeout messages received where
-    /// N=timeout_quorum_size.  The round of PacemakerTimeoutCertificate is determined as
+    /// N=timeout_quorum_voting_power.  The round of PacemakerTimeoutCertificate is determined as
     /// the smallest of round of all messages used to generate this certificate.
     ///
-    /// For example, if timeout_certificate_quorum_size=3 and we received unique author timeouts
-    /// for rounds (1,2,3,4), then rounds (2,3,4) would form PacemakerTimeoutCertificate with
-    /// round=2.
+    /// For example, if timeout_certificate quorum_voting_power=3 and we received unique author
+    /// timeouts with equal voting power for rounds (1,2,3,4), then rounds (2,3,4) would form
+    /// PacemakerTimeoutCertificate with round=2.
     fn generate_timeout_certificate(
         author_to_received_timeouts: &HashMap<Author, PacemakerTimeout>,
-        quorum_size: usize,
+        validator_verifier: Arc<ValidatorVerifier>,
     ) -> Option<PacemakerTimeoutCertificate> {
-        if author_to_received_timeouts.values().len() < quorum_size {
-            return None;
+        // Sort the timeouts by round, highest to lowest and then, in order, aggregate
+        // enough voting power to assemble a PacemakerTimeoutCertificate if possible.
+        let mut received_timeouts: Vec<PacemakerTimeout> =
+            author_to_received_timeouts.values().cloned().collect();
+        received_timeouts.sort_by(|x, y| y.round().cmp(&x.round()));
+
+        let mut combined_voting_power = 0;
+        for (i, received_timeout) in received_timeouts.iter().enumerate() {
+            match validator_verifier.get_voting_power(&received_timeout.author()) {
+                Some(voting_power) => combined_voting_power += voting_power,
+                None => panic!(
+                    "Illegal that author {} is not part of validator set",
+                    received_timeout.author()
+                ),
+            }
+            if combined_voting_power >= validator_verifier.quorum_voting_power() {
+                return Some(PacemakerTimeoutCertificate::new(
+                    received_timeout.round(),
+                    Vec::from_iter(received_timeouts[..i].iter().cloned()),
+                ));
+            }
```

Comment:
Even if you're intent on the `for`, you could write this, it's the same idea as what I expressed with a `take_while` in [this older comment](https://github.com/libra/libra/pull/1149#discussion_r330382605):
```
        let mut combined_voting_power = 0;
        let mut combined_received_timeouts = vec![];
        for received_timeout in received_timeouts.iter() {
            match validator_verifier.get_voting_power(&received_timeout.author()) {
                Some(voting_power) => {
                    combined_received_timeouts.push(received_timeout.clone());
                    combined_voting_power += voting_power
                }
                None => panic!(
                    "Illegal that author {} is not part of validator set",
                    received_timeout.author()
                ),
            }
            if combined_voting_power >= validator_verifier.quorum_voting_power() {
                return Some(PacemakerTimeoutCertificate::new(
                    received_timeout.round(),
                    combined_received_timeouts,
                ));
            }
        }
```

Reply:
Removing the index is a bit nicer - even at the cost of a few more lines.  Changing.